### ISSUE TITLE
Not able to start in tomcat 7

### ISSUE THREAD

**Issue by jenshp** [moqui/moqui-framework#91]
The feature of looking for the ServerContainer triggers a casting Exception (see below). This does not happen in tomcat 8. Is this a problem with Tomcat, or the servlet spec used (3.0 for tomcat 7, 3.1 for tomcat 8)? Does this mean moqui-2.0 will not work with tomcat 7?

Aug 18, 2016 2:17:24 PM org.apache.catalina.core.StandardContext listenerStart
SEVERE: Exception sending context initialized event to listener instance of class org.moqui.impl.webapp.MoquiContextListener
org.codehaus.groovy.runtime.typehandling.GroovyCastException: Cannot cast object 'org.apache.tomcat.websocket.server.WsServerContainer@1344f181' with class 'org.apache.tomcat.websocket.server.WsServerContainer' to class 'javax.websocket.server.ServerContainer'
        at org.codehaus.groovy.runtime.typehandling.DefaultTypeTransformation.continueCastOnSAM(DefaultTypeTransformation.java:405)
        at org.codehaus.groovy.runtime.typehandling.DefaultTypeTransformation.continueCastOnNumber(DefaultTypeTransformation.java:319)
        at org.codehaus.groovy.runtime.typehandling.DefaultTypeTransformation.castToType(DefaultTypeTransformation.java:232)
        at org.codehaus.groovy.runtime.ScriptBytecodeAdapter.castToType(ScriptBytecodeAdapter.java:603)
        at org.moqui.impl.context.ExecutionContextFactoryImpl.initServletContext(ExecutionContextFactoryImpl.groovy:1019)
        at org.moqui.impl.webapp.MoquiContextListener.contextInitialized(MoquiContextListener.groovy:81)
        at org.apache.catalina.core.StandardContext.listenerStart(StandardContext.java:4973)
        at org.apache.catalina.core.StandardContext.startInternal(StandardContext.java:5467)
        at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:150)
        at org.apache.catalina.core.ContainerBase.addChildInternal(ContainerBase.java:901)
        at org.apache.catalina.core.ContainerBase.addChild(ContainerBase.java:877)
        at org.apache.catalina.core.StandardHost.addChild(StandardHost.java:632)
        at org.apache.catalina.startup.HostConfig.deployWAR(HostConfig.java:1083)
        at org.apache.catalina.startup.HostConfig$DeployWar.run(HostConfig.java:1880)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)

**Comment by chunlinyao** [comment]
I think `javax.websocket.server.ServerContainer` is loaded by multiple ClassLoaders. We shipped a  websocket api jar file in `WEB-INF/lib`. Tomcat also include a websocket api jar. I don't known why tomcat8 can work. Can you remove `WEB-INF/lib/javax.websocket-api-1.0.jar` from war file and deploy that war file again.

**Comment by jonesde** [comment]
Aside from the Websocket API (not sure why that is having a binding issue with Tomcat 7...) you'll likely have issues with other things once you're past that (i.e. if removing javax.websocket-api-1.0.jar works). Like you mentioned, Moqui is using the Servlet 3.1 spec now.

In general with the current code base and the 2.0.0 release we're finally moving to some newer stuff like requiring Java 8 and more recent servlet containers too. There was never any official support for external servlet containers, but I generally test with at least Tomcat and Jetty now and it should work on any that support the spec versions used in Moqui (which you can see most easily in the framework/build.gradle file with the javax spec dependencies).

In short no, Tomcat 7 is no longer supported just like Java 7 is no longer supported since the 1.6.2 release (along with various other deprecations and updates in the upcoming 2.0.0 release).

**Comment by chunlinyao** [comment]
FYI: The reason tomcat 8 not throw this exception is because `WebappClassLoader`  in tomcat8 filter out `javax.websocket`.

**Comment by jenshp** [comment]
@chunlinyao it did not work by removing the WEB-INF/lib/javax.websocket-api-1.0.jar from the WAR file, but it did work by configuring tomcat 7 classloader with delegate=true for the context in which moqui is loaded.
However, it seems to be a fragile fix, as moqui will be using the classes provided by the server instead of the ones shipped in the WAR.

**Comment by jonesde** [comment]
For things like spec jars Moqui really doesn't care which one is used, and it should even work fine with a Websocket 1.1 spec jar instead of the 1.0 (if that is what the underlying websocket implementation supports).
